src/mscorlib/src/System/Number.cs

```diff
@@ -721,6 +721,78 @@ internal unsafe static Double ParseDouble(String value, NumberStyles options, Nu
             return d;
         }
 
+        internal unsafe static Int32 ParseInt32Fast(String s, NumberFormatInfo info)
+        {
+            if (s == null)
+            {
+                throw new ArgumentNullException(nameof(String));
+            }
+
+            const int StateSign = 1 << 0;
+            const int StateNegative = 1 << 1;
+            const int StateDigits = 1 << 2;
+
+            var state = 0;
+            var ret = 0;
+            fixed (char* sptr = s)
+            {
+                var len = s.Length * 2;
+                var cptr = sptr;
+                while (len > 0)
+                {
+                    len -= 2;
+                    var c = *cptr;
+                    if (c >= '0' && c <= '9' && (state & StateDigits) == 0)
+                    {
+                        state |= StateSign;
+                        ret = ret * 10 + (c - '0');
+                    }
+                    else if ((state & StateSign) == 0 && MatchChars(cptr, info.positiveSign) != null)
+                    {
+                        state |= StateSign;
+                    }
+                    else if ((state & StateSign) == 0 && MatchChars(cptr, info.negativeSign) != null)
+                    {
+                        state |= StateSign | StateNegative;
+                    }
+                    // leading whitespaces
+                    else if (IsWhite(c) && (state & StateSign) == 0)
+                    {
+                    }
+                    // trailing whitespaces
+                    else if (IsWhite(c))
+                    {
+                        state |= StateDigits;
+                    }
+                    else
+                    {
+                        throw new OverflowException(SR.Format_InvalidString);
+                    }
+
+                    cptr++;
+                }
+            }
+
+
+            if ((state & StateNegative) > 0)
+            {
+                ret = -ret;
+                if (ret > 0)
```

Comment:
I don't think these 2 checks are enough to detect overflow. They're enough only if we limit the number of digits you parse and I don't see such a limit in your code.